File: 0198-house-robber/0198-house-robber.java
class Solution {
    int[][] dp = new int[400][2];
    boolean[][] visited = new boolean[400][2];

    public int rob(int[] nums) {
        return maximumProfit(nums, 0, false);
    }

    private int maximumProfit(int[] nums, int i, boolean previousWasTaken) {
        if (i == nums.length) {
            return 0;
        }

        if (visited[i][previousWasTaken ? 1: 0]) {
            return dp[i][previousWasTaken ? 1 : 0];
        }

        int profit = 0;
        visited[i][previousWasTaken ? 1: 0] = true;

        if (!previousWasTaken) {
            profit = nums[i] + maximumProfit(nums, i + 1, true);
        }
        profit = Math.max(profit, maximumProfit(nums, i + 1, false));
        dp[i][previousWasTaken ? 1 : 0] = profit;

        return profit;
    }
}

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = {}

        def solve(i: int):
            if i >= len(nums):
                return 0

            if i in dp:
                return dp[i]

            dp[i] = max(solve(i + 1), nums[i] + solve(i + 2))

            return dp[i]

        return solve(0)
